File: assemblyscript/assembly/wire.ts
// The wire codec primitives, byte-for-byte compatible with the Rust
// `propify-sandbox-abi` codec.
//
// Everything is little-endian. wasm linear memory is little-endian and AS
// `load<T>`/`store<T>` are little-endian, so a multi-byte `store<i32>`/`store<i64>`
// at any offset writes the exact LE bytes the Rust `to_le_bytes` encoder produces.
// wasm permits unaligned accesses with identical results to aligned ones, so the
// variable-length `String` fields (which push later fields off natural alignment)
// are still encoded and decoded correctly.
//
// We work at the raw-pointer level (`load<u8>`/`store<u8>` and friends) rather than
// through the AS `String`/`Array` types on purpose: it keeps the import surface to
// exactly the four `propify` functions (the `String` type can pull UTF-8 machinery,
// and through it `abort`), and it lets a decoded asset alias the source buffer with
// zero copying.

import { Decimal } from "./decimal";

/**
 * A growable byte buffer in the guest's own linear memory, used to encode an
 * outgoing message.
 *
 * Backed by `heap.alloc`; it grows by allocating a larger block and copying. Under
 * the `stub` runtime `heap.free` is a no-op, but every block is still released via
 * [`Writer#free`] so the alloc/dealloc pairing is explicit and the code stays
 * correct under any allocator.
 */
export class Writer {
  private ptr: usize;
  private cap: i32;
  private len: i32;

  constructor(initialCapacity: i32) {
    this.cap = initialCapacity > 0 ? initialCapacity : 16;
    this.ptr = heap.alloc(<usize>this.cap);
    this.len = 0;
  }

  /** Number of bytes written so far. */
  get length(): i32 {
    return this.len;
  }

  /** Base offset of the encoded bytes in linear memory. */
  get pointer(): usize {
    return this.ptr;
  }

  /** Releases the backing block. Pairs with the constructor's allocation. */
  free(): void {
    heap.free(this.ptr);
  }

  private ensure(extra: i32): void {
    const need = this.len + extra;
    if (need <= this.cap) return;
    let newCap = this.cap * 2;
    while (newCap < need) newCap *= 2;
    const newPtr = heap.alloc(<usize>newCap);
    memory.copy(newPtr, this.ptr, <usize>this.len);
    heap.free(this.ptr);
    this.ptr = newPtr;
    this.cap = newCap;
  }

  putU8(value: u8): void {
    this.ensure(1);
    store<u8>(this.ptr + <usize>this.len, value);
    this.len += 1;
  }

  putU32(value: u32): void {
    this.ensure(4);
    store<u32>(this.ptr + <usize>this.len, value);
    this.len += 4;
  }

  putI32(value: i32): void {
    this.ensure(4);
    store<i32>(this.ptr + <usize>this.len, value);
    this.len += 4;
  }

  putI64(value: i64): void {
    this.ensure(8);
    store<i64>(this.ptr + <usize>this.len, value);
    this.len += 8;
  }

  putBool(value: bool): void {
    this.putU8(value ? <u8>1 : <u8>0);
  }

  /** Copies `len` raw bytes from `srcPtr` into the buffer. */
  putBytes(srcPtr: usize, len: i32): void {
    if (len <= 0) return;
    this.ensure(len);
    memory.copy(this.ptr + <usize>this.len, srcPtr, <usize>len);
    this.len += len;
  }

  /** Encodes a `String` field: a u32 LE byte length, then the raw bytes. */
  putString(srcPtr: usize, len: i32): void {
    this.putU32(<u32>len);
    this.putBytes(srcPtr, len);
  }

  /** Encodes a `Decimal`: i128 mantissa (16 LE, two i64 halves) + i32 scale (4 LE). */
  putDecimal(value: Decimal): void {
    this.putI64(value.mantissaLow);
    this.putI64(value.mantissaHigh);
    this.putI32(value.scale);
  }

  /** Encodes an `Option<Decimal>`: tag 0 = None, tag 1 = Some + the decimal. */
  putOptionDecimal(value: Decimal | null): void {
    if (value === null) {
      this.putU8(0);
    } else {
      this.putU8(1);
      this.putDecimal(value);
    }
  }
}

/**
 * A bounds-checked, forward-only reader over a byte range in linear memory.
 *
 * Mirrors the Rust codec's `Cursor`: every read first checks that enough bytes
 * remain and flips [`Reader#failed`] instead of reading out of bounds. A caller
 * inspects `failed` after a sequence of reads to decide whether the decode held.
 * The source bytes are the host-written input buffer, which the SDK keeps alive for
 * the whole tick, so decoded byte slices may alias it safely.
 */
export class Reader {
  private base: usize;
  private end: usize;
  private pos: usize;
  /** Set once any read ran past the end; the decoded value is then meaningless. */
  failed: bool;

  constructor(ptr: usize, len: i32) {
    const safeLen: i32 = len > 0 ? len : 0;
    this.base = ptr;
    this.end = ptr + <usize>safeLen;
    this.pos = ptr;
    this.failed = false;
  }

  private has(n: usize): bool {
    return !this.failed && this.end - this.pos >= n;
  }

  /**
   * The reader's current absolute pointer into linear memory. Used by
   * [`MarketWindow#decode`] to record where the candle array begins so an individual
   * candle can be read in place (copy-free) on demand.
   */
  get position(): usize {
    return this.pos;
  }

  /** Bytes left to read between the cursor and the end of the buffer. */
  get remaining(): i32 {
    return <i32>(this.end - this.pos);
  }

  readU8(): u8 {
    if (!this.has(1)) {
      this.failed = true;
      return 0;
    }
    const v = load<u8>(this.pos);
    this.pos += 1;
    return v;
  }

  readU32(): u32 {
    if (!this.has(4)) {
      this.failed = true;
      return 0;
    }
    const v = load<u32>(this.pos);
    this.pos += 4;
    return v;
  }

  readI32(): i32 {
    if (!this.has(4)) {
      this.failed = true;
      return 0;
    }
    const v = load<i32>(this.pos);
    this.pos += 4;
    return v;
  }

  readI64(): i64 {
    if (!this.has(8)) {
      this.failed = true;
      return 0;
    }
    const v = load<i64>(this.pos);
    this.pos += 8;
    return v;
  }

  readBool(): bool {
    return this.readU8() != 0;
  }

  /**
   * Reads a `String` field and returns its byte range as a [`ByteSlice`] aliasing
   * the source buffer (no copy). The length prefix is validated against the
   * remaining bytes.
   */
  readString(): ByteSlice {
    const len = this.readU32();
    if (this.failed || !this.has(<usize>len)) {
      this.failed = true;
      return new ByteSlice(0, 0);
    }
    const slice = new ByteSlice(this.pos, <i32>len);
    this.pos += <usize>len;
    return slice;
  }

  /** Reads a `Decimal`: i128 mantissa (16 LE, two i64 halves) + i32 scale (4 LE). */
  readDecimal(): Decimal {
    const low = this.readI64();
    const high = this.readI64();
    const scale = this.readI32();
    return new Decimal(low, high, scale);
  }

  /** Advances over a `Decimal` (20 bytes) without materializing it. */
  skipDecimal(): void {
    this.readI64();
    this.readI64();
    this.readI32();
  }
}

/**
 * A borrowed view of bytes in linear memory: a base pointer and a length.
 *
 * Used for `String` fields so an asset symbol can be passed straight through from a
 * decoded `MarketSnapshot` into an outgoing `OrderIntentBody` without ever building
 * an AS `String`.
 */
export class ByteSlice {
  constructor(public ptr: usize, public len: i32) {}
}

/**
 * Byte-exact equality between a [`ByteSlice`] and an ASCII literal held in a
 * `StaticArray<u8>`, used to look up a strategy parameter by name without UTF-8
 * decoding either side.
 */
export function sliceEqualsAscii(slice: ByteSlice, ascii: StaticArray<u8>): bool {
  const len = ascii.length;
  if (slice.len != len) return false;
  for (let i = 0; i < len; i++) {
    if (load<u8>(slice.ptr + <usize>i) != unchecked(ascii[i])) return false;
  }
  return true;
}

File: tinygo/propify/wire.go
// The wire codec primitives, byte-for-byte compatible with the Rust
// `propify-sandbox-abi` codec and the AssemblyScript SDK.
//
// Everything is little-endian, two's complement. We do every multi-byte read and
// write with MANUAL little-endian byte shifts rather than `encoding/binary`, because
// that package is marked failing under TinyGo and pulling it in could also widen the
// import surface. Manual shifts keep the dependency set to nothing and produce the
// exact bytes the Rust `to_le_bytes` encoder produces.
//
// We work over plain `[]byte` slices into the guest's own linear memory. A decoded
// `String` field is returned as a sub-slice that ALIASES the source buffer (no copy),
// so a bot can pass an asset symbol straight through from a decoded snapshot into an
// outgoing order. The host keeps the input buffer alive for the whole tick, so the
// alias is valid until the tick ends.
package propify

// Writer is an append-only byte buffer used to encode one outgoing message. It is backed
// by a fixed, full-length region of the guest's bump arena (see newWriter), sized to the
// host's MAX_MESSAGE_BYTES, and writes through a manual position cursor with plain indexed
// stores. It deliberately avoids Go's `append`, which lowers to `runtime.sliceAppend` and
// can call `runtime.alloc` — and any heap allocation traps on this target because the host
// never calls `_initialize` to bootstrap the allocator. The cursor never exceeds the
// reserved length, so no store ever needs to grow. The caller pins the finished slice into
// linear memory (see emitIntent) before handing its pointer to the host.
type Writer struct {
	buf []byte
	pos int
}

// newWriter returns a Writer over a full-length, max-message-sized arena reservation, with
// the cursor at zero. Stores index into that fixed region, so no encode ever reaches the Go
// heap. If the arena cannot satisfy the reservation the buffer is nil (length 0); every
// store then becomes a bounds-guarded no-op and emitIntent offers an empty message, which
// the host rejects — a safe degradation rather than a heap trap. In practice the arena is
// sized so this never happens.
func newWriter() Writer {
	_, region := allocBuffer(maxMessageBytes)
	return Writer{buf: region}
}

// Bytes returns the encoded bytes written so far (the reserved region sliced to the cursor).
func (w *Writer) Bytes() []byte {
	return w.buf[:w.pos]
}

// putByte writes one byte at the cursor and advances it. It is bounds-guarded: a write past
// the reserved region is dropped rather than allowed to panic (which would trap). The region
// is sized to the message cap, so this guard never fires for a valid message.
func (w *Writer) putByte(v uint8) {
	if w.pos < len(w.buf) {
		w.buf[w.pos] = v
		w.pos++
	}
}

// PutU8 writes a single byte.
func (w *Writer) PutU8(v uint8) {
	w.putByte(v)
}

// PutU32 writes a u32 in little-endian order.
func (w *Writer) PutU32(v uint32) {
	w.putByte(byte(v))
	w.putByte(byte(v >> 8))
	w.putByte(byte(v >> 16))
	w.putByte(byte(v >> 24))
}

// PutI32 writes an i32 in little-endian, two's-complement order.
func (w *Writer) PutI32(v int32) {
	w.PutU32(uint32(v))
}

// PutU64 writes a u64 in little-endian order.
func (w *Writer) PutU64(v uint64) {
	w.putByte(byte(v))
	w.putByte(byte(v >> 8))
	w.putByte(byte(v >> 16))
	w.putByte(byte(v >> 24))
	w.putByte(byte(v >> 32))
	w.putByte(byte(v >> 40))
	w.putByte(byte(v >> 48))
	w.putByte(byte(v >> 56))
}

// PutI64 writes an i64 in little-endian, two's-complement order.
func (w *Writer) PutI64(v int64) {
	w.PutU64(uint64(v))
}

// PutBool writes a bool as 1 (true) or 0 (false), matching the Rust codec.
func (w *Writer) PutBool(v bool) {
	if v {
		w.putByte(1)
	} else {
		w.putByte(0)
	}
}

// PutBytes writes raw bytes verbatim through the cursor.
func (w *Writer) PutBytes(b []byte) {
	for i := 0; i < len(b); i++ {
		w.putByte(b[i])
	}
}

// PutString encodes a String field: a u32 little-endian byte length, then the bytes.
func (w *Writer) PutString(b []byte) {
	w.PutU32(uint32(len(b)))
	w.PutBytes(b)
}

// PutDecimal encodes a Decimal: i128 mantissa (16 LE, two u64 halves) then i32 scale
// (4 LE). Low half first, high half second, matching the i128 little-endian layout.
func (w *Writer) PutDecimal(d Decimal) {
	w.PutU64(d.MantissaLow)
	w.PutU64(d.MantissaHigh)
	w.PutI32(d.Scale)
}

// PutOptionDecimal encodes an Option<Decimal>: tag 0 = None, tag 1 = Some + the
// decimal. A nil pointer is None.
func (w *Writer) PutOptionDecimal(d *Decimal) {
	if d == nil {
		w.PutU8(0)
		return
	}
	w.PutU8(1)
	w.PutDecimal(*d)
}

// Reader is a bounds-checked, forward-only reader over a byte slice in linear memory.
//
// It mirrors the Rust codec's `Cursor` and the AssemblyScript `Reader`: every read
// first checks that enough bytes remain and flips `failed` instead of reading out of
// bounds. A caller inspects `failed` after a sequence of reads to decide whether the
// decode held. Once `failed` is set, later reads return zero values and stay failed.
type Reader struct {
	buf    []byte
	pos    int
	failed bool
}

// sharedReader is the single Reader instance every decode reuses. Returning the address
// of a freshly built `&Reader{}` would make it escape onto the Go heap — TinyGo's escape
// analysis cannot prove the pointer stays local once it is passed to the pointer-receiver
// read methods — and any heap allocation traps on this target because the host never calls
// `_initialize`. A package-level static lives in the data/bss segment instead, needing no
// allocation and no runtime init.
//
// One shared instance is safe here: the guest is single-threaded and every decode is
// strictly sequential and fully consumed before the next begins (the market snapshot, then
// the account view, then each strategy-parameter lookup), so no two readers are ever live
// at once. Each NewReader call rebinds the buffer and resets the cursor and failed flag.
var sharedReader Reader

// NewReader binds the shared Reader to the given buffer (resetting its cursor and failed
// flag) and returns it. See sharedReader for why a single static instance is used rather
// than a heap allocation.
func NewReader(buf []byte) *Reader {
	sharedReader = Reader{buf: buf}
	return &sharedReader
}

// has reports whether at least n more bytes are available and the reader has not
// already failed. A negative n (from an out-of-range length prefix on a 32-bit guest)
// is treated as "not available" so a hostile length cannot drive an out-of-bounds
// slice.
func (r *Reader) has(n int) bool {
	return !r.failed && n >= 0 && len(r.buf)-r.pos >= n
}

func (r *Reader) readU8() uint8 {
	if !r.has(1) {
		r.failed = true
		return 0
	}
	v := r.buf[r.pos]
	r.pos++
	return v
}

func (r *Reader) readU32() uint32 {
	if !r.has(4) {
		r.failed = true
		return 0
	}
	b := r.buf[r.pos:]
	v := uint32(b[0]) | uint32(b[1])<<8 | uint32(b[2])<<16 | uint32(b[3])<<24
	r.pos += 4
	return v
}

func (r *Reader) readI32() int32 {
	return int32(r.readU32())
}

func (r *Reader) readU64() uint64 {
	if !r.has(8) {
		r.failed = true
		return 0
	}
	b := r.buf[r.pos:]
	v := uint64(b[0]) | uint64(b[1])<<8 | uint64(b[2])<<16 | uint64(b[3])<<24 |
		uint64(b[4])<<32 | uint64(b[5])<<40 | uint64(b[6])<<48 | uint64(b[7])<<56
	r.pos += 8
	return v
}

func (r *Reader) readI64() int64 {
	return int64(r.readU64())
}

func (r *Reader) readBool() bool {
	return r.readU8() != 0
}

// readString reads a String field and returns its bytes as a sub-slice aliasing the
// source buffer (no copy). The length prefix is validated against the remaining
// bytes; an over-long prefix fails the reader and returns nil.
func (r *Reader) readString() []byte {
	n := int(r.readU32())
	if r.failed || !r.has(n) {
		r.failed = true
		return nil
	}
	s := r.buf[r.pos : r.pos+n]
	r.pos += n
	return s
}

// readDecimal reads a Decimal: i128 mantissa (16 LE, two u64 halves) then i32 scale
// (4 LE).
func (r *Reader) readDecimal() Decimal {
	low := r.readU64()
	high := r.readU64()
	scale := r.readI32()
	return Decimal{MantissaLow: low, MantissaHigh: high, Scale: scale}
}

// bytesEqual reports byte-exact equality between two slices, used to look up a
// strategy parameter by name without any UTF-8 decoding on either side.
func bytesEqual(a, b []byte) bool {
	if len(a) != len(b) {
		return false
	}
	for i := range a {
		if a[i] != b[i] {
			return false
		}
	}
	return true
}
